source/connection-monitoring-and-pooling/tests/logging/connection-logging.yml

```diff
@@ -0,0 +1,194 @@
+            component: connection
+            data:
+              message: "Connection pool ready"
+              serverHost: { $$type: string }
+              serverPort: { $$type: [int, long] }
+
+          - level: debug
+            component: connection
+            data:
+              message: "Connection checkout started"
+              serverHost: { $$type: string }
+              serverPort: { $$type: [int, long] }
+
+          - level: debug
+            component: connection
+            data:
+              message: "Connection created"
+              driverConnectionId: { $$type: [int, long] }
+              serverHost: { $$type: string }
+              serverPort: { $$type: [int, long] }
+
+          - level: debug
+            component: connection
+            data:
+              message: "Connection ready"
+              driverConnectionId: { $$type: [int, long] }
+              serverHost: { $$type: string }
+              serverPort: { $$type: [int, long] }
+          
+          - level: debug
+            component: connection
+            data:
+              message: "Connection checked out"
+              driverConnectionId: { $$type: [int, long] }
+              serverHost: { $$type: string }
+              serverPort: { $$type: [int, long] }
+
+          - level: debug
+            component: connection
+            data:
+              message: "Connection checked in"
+              driverConnectionId: { $$type: [int, long] }
+              serverHost: { $$type: string }
+              serverPort: { $$type: [int, long] }
+
+          - level: debug
+            component: connection
+            data:
+              message: "Connection closed"
+              driverConnectionId: { $$type: [int, long] }
+              serverHost: { $$type: string }
+              serverPort: { $$type: [int, long] }
+              reason: "Connection pool was closed"
+
+          - level: debug
+            component: connection
+            data:
+              message: "Connection pool closed"
+              serverHost: { $$type: string }
+              serverPort: { $$type: [int, long] }
+
+  # This test exists to probide coverage of checkout failed and pool cleared events.
+  - description: "Connection checkout fails due to error establishing connection"
+    runOnRequirements:
+      - auth: true
+        minServerVersion: "4.0" # failCommand was added to mongod in 4.0
+    operations:
+      - name: createEntities
+        object: testRunner
+        arguments:
+          entities:
+            - client:
+                id: &client client
+                uriOptions:
+                  retryReads: false
+                  appname: &clientAppName clientAppName
+                  # use a high heartbeatFrequencyMS to avoid a successful monitor check marking the pool as
+                  # ready (and emitting another event) during the course of test execution.
+                  heartbeatFrequencyMS: 10000
```

Comment:
at least in Rust I was running into a failure on server versions that do not support streamable monitoring (< 4.4) like the following:
* failpoint is hit: connection pool is cleared, server is marked unknown, in progress check is canceled 
* the next check starts and succeeds quickly so pool is re-marked ready, log message is emitted
* we assert on emitted log messages and there is an extra "pool ready" message at the end 

I _think_ this wasn't happening with the streaming protocol because the server waits the full heartbeatFrequencyMS/maxAwaitTimeMS to send an update in response to the check following the failpoint rather than returning immediately, since from its perspective nothing actually changed, so in that case the test gets a little extra time to make the assertions before the pool is marked ready again and a new event goes out. (does this sound right @patrickfreed?) 
that said, it seems possible to hit this issue with the streaming protocol too with a low enough heartbeatFrequencyMS + sufficient delay in asserting on emitted log messages.

I did not have a strong reason for 10 seconds specifically here; I just chose a large value that seemed like it would very safely allow the test to complete before a second heartbeat after the initial one which discovers the server succeeds.

Reply:
> I think this wasn't happening with the streaming protocol because the server waits the full heartbeatFrequencyMS/maxAwaitTimeMS to send an update in response to the check following the failpoint rather than returning immediately

I'm actually not sure if this hypothesis makes sense because I think upon seeing the network error the server close the monitoring connection and open a new one on which it will re-negotiate the monitoring protocol. and so the first check following the failpoint would not be awaitable. I'll try to figure out what's actually happening tomorrow 🙂

Reply:
I figured out what was going on here. some key observations:
1) the Rust test runner coincidentally waits for exactly 500ms=default heartbeatFrequencyMS before making assertions on events/log messages, in order to give the events time to come in.
2) after an in-progress streaming monitoring check is canceled due to a network error during handshake, we wait heartbeatFrequencyMS post cancellation before starting a new one. 
3) however, if a network error during handshake occurs between heartbeats in the polling protocol, it has no impact on the timing of the next heartbeat. 

given those, what was happening in the polling protocol was something like:
1. t=0 heartbeat 1 starts
2. t=2 heartbeat 1 succeeds; server marked standalone 
3. t=10 failpoint hit; test runner begins 500ms sleep
4. t=15 connection pool is cleared and server is marked unknown
5. t=502 heartbeat 2 begins (per observation 3, 500ms after last heartbeat)
6. t=504 heartbeat 2 succeeds
7. t=508 server marked as standalone and connection pool marked ready
8. 7=510 test runner wakes up and starts asserting on events

vs in the streaming protocol:
1. t=0 heartbeat 1 starts (first one is non awaitable)
2. t=2 heartbeat 1 succeeds; server marked standalone 
3. t=2 heartbeat 2 begins (awaitable)
4. t=10 failpoint hit; test runner begins 500ms sleep
5. t=15 connection pool is cleared and server is marked unknown
6. t=15 in-progress heartbeat is canceled
7. t=510 test runner wakes up and asserts on events
8. t=515 heartbeat 3 begins (500ms after cancellation, due to observation 2 above) (non awaitable due to coming after the failpoint error) 

If the test runner sleep is increased to 510 seconds, the test also starts to fail in the streaming protocol, since it gives enough time for heartbeat 3 to happen. 

the Rust test runner's time spent sleeping could possibly be shortened, but even so and especially given that drivers tend to decrease heartbeatFrequencyMS to speed up tests it seems like others could run into this if their assertions aren't completed before the next heartbeat. given this I think it still makes sense to keep the high value here to remove any chance that a new heartbeat completes before the test completes.

Reply:
Sounds good to me! great sleuthing to figure out the timing issue